src/Domain/DonationRepository.php

```diff
@@ -455,6 +455,78 @@ public function findWithTransferIdInArray(array $transferIds): array
         return $donations;
     }
 
+    /**
+     * Taking the floor of the current minute as N and looking between N-16 minutes and
+     * N-1 minutes for donations with >£0 matching assigned, returns:
+     * * if there are less than 20 such donations, null; or
+     * * if there are 20+ such donations, the ratio of those which are complete.
+     */
+    public function getRecentHighVolumeCompletionRatio(\DateTimeImmutable $startOfThisMinute): ?float
+    {
+        $oneMinuteAgo = $startOfThisMinute->sub(new \DateInterval('PT1M'));
+        $sixteenMinutesAgo = $startOfThisMinute->sub(new \DateInterval('PT16M'));
+
+        $qb = $this->getEntityManager()->createQueryBuilder()
+            ->select(
+                'COUNT(d.id) as donationCount, SUM(CASE WHEN d.donationStatus IN ' .
+                '(:completeStatuses) THEN 1 ELSE 0 END) as completeCount'
+            )
+            ->from(Donation::class, 'd')
+            ->leftJoin('d.fundingWithdrawals', 'fw')
+            ->where('d.createdAt >= :start')
+            ->andWhere('d.createdAt < :end')
+            ->having('SUM(fw.amount) > 0')
+            ->setParameter('start', $sixteenMinutesAgo)
+            ->setParameter('end', $oneMinuteAgo)
+            ->setParameter(
+                'completeStatuses',
+                array_map(static fn(DonationStatus $s) => $s->value, DonationStatus::SUCCESS_STATUSES),
+            );
+
+        /**
+         * @var array{donationCount: int, completeCount: int}|null $result
+         */
+        $result = $qb->getQuery()->getOneOrNullResult(Query::HYDRATE_ARRAY);
```

Comment:
I do still think DQL is quite a bit easier to read and edit than the query builder, and is well supported by PHPStorm. We lose out a bit on Psalm checks but I don't think Psalm does enough to be worthwhile. I converted it here by adding var dumping the output of `$qb->getQuery()->getDql()` and then reformatting by hand:

```suggestion
        $query = $this->getEntityManager()->createQuery(<<<'DQL'
            SELECT
            COUNT(d.id) as donationCount, 
            SUM(CASE WHEN d.donationStatus IN (:completeStatuses) THEN 1 ELSE 0 END) as completeCount
            FROM MatchBot\Domain\Donation d 
            LEFT JOIN d.fundingWithdrawals fw
            WHERE d.createdAt >= :start
            AND d.createdAt < :end HAVING SUM(fw.amount) > 0
        DQL
        );
        $query->setParameter('start', $sixteenMinutesAgo);
        $query->setParameter('end', $oneMinuteAgo);

        /**
         * @var array{donationCount: int, completeCount: int}|null $result
         */
        $result = $query->getOneOrNullResult(Query::HYDRATE_ARRAY);
```

Reply:
Same applies to other uses of `createQueryBuilder` below.

Reply:
It's true that I'm getting completion suggestions with either option. I do think there are stronger checks when building the query in code, e.g. getting clear errors about entity classes that don't exist, without waiting until runtime. But have changed all the new queries to DQL for now.

Reply:
Yep. We could get https://github.com/phpstan/phpstan-doctrine to do static checks for DQL. But I don't think it's hugely important, we only have to run it once to catch runtime errors. It's not like we'd have to run multiple scenarios for different code paths.